netket/utils/group/permutation_group.py

```diff
@@ -0,0 +1,203 @@
+        """
+        Creates a `Permutation` from an array of preimages of :code:`range(N)`
+
+        Arguments:
+            permutation: a 1D array listing :math:`g^{-1}(x)` for all :math:`0\le x < N` (i.e., `V[permutation]` permutes the elements of `V` as desired)
+            name: optional, custom name for the permutation
+
+        Returns:
+            a `Permutation` object encoding the same permutation
+        """
+        self.permutation = HashableArray(np.asarray(permutation))
+        self.__name = name
+
+    def __call__(self, x):
+        return x[..., self.permutation]
+
+    def __hash__(self):
+        return hash(self.permutation)
+
+    def __eq__(self, other):
+        if isinstance(other, Permutation):
+            return self.permutation == other.permutation
+        else:
+            return False
+
+    @property
+    def _name(self):
+        return self.__name
+
+    def __repr__(self):
+        if self._name is not None:
+            return self._name
+        else:
+            return f"Permutation({np.asarray(self).tolist()})"
+
+    def __array__(self, dtype: DType = None):
+        return np.asarray(self.permutation, dtype)
+
+
+@dispatch
+def product(p: Permutation, q: Permutation):
+    name = None if p._name is None and q._name is None else f"{p} @ {q}"
+    return Permutation(p(np.asarray(q)), name)
+
+
+@struct.dataclass
+class PermutationGroup(Group):
+    """
+    Collection of permutation operations acting on sequences of length :code:`degree`.
+
+    Group elements need not all be of type :ref:`netket.utils.symmetry.Permutation`,
+    only act as such on a sequence when called. Currently, however, only `Identity`
+    and `Permutation` have canonical forms implemented.
+
+    The class can contain elements that are distinct as objects (e.g.,
+    :code:`Identity()` and :code:`Translation((0,))`) but have identical action.
+    Those can be removed by calling :code:`remove_duplicates`.
+    """
+
+    degree: int
+    """Number of elements the permutations act on."""
+
+    def __hash__(self):
+        return super().__hash__()
+
+    def __matmul__(self, other) -> "PermutationGroup":
+        if not isinstance(other, PermutationGroup):
+            raise ValueError(
+                "Incompatible groups (`PermutationGroup` and something else)"
+            )
+        elif self.degree != other.degree:
+            raise ValueError(
+                "Incompatible groups (`PermutationGroup`s of different degree)"
+            )
+
+        return PermutationGroup(super().__matmul__(other).elems, self.degree)
+
+    def _canonical(self, x: Element) -> Array:
+        if isinstance(x, Identity):
+            return np.arange(self.degree, dtype=int)
+        elif isinstance(x, Permutation):
+            return np.asarray(x.permutation)
+        else:
+            raise ValueError(
+                "`PermutationGroup` only supports `Identity` and `Permutation` elements"
+            )
```

Comment:
Isn't this just a less general way of doing
```python
x(np.arange(self.degree, dtype=int))
```
which works for any type of `Element` with compatible action?

Reply:
```suggestion
        return x(np.arange(self.degree, dtype=int))
```

Reply:
I'll check if this might break anything somewhere in the implementation before committing, but it looks sound.